### PR TITLE
Perform an inversion of control in hint validation for eapply.

### PR DESCRIPTION
**Pull request by ppedrot**
We move the "verbose" case to the only point it is actually used. It is a bit unfortunate since it implies a bit of code duplication, but this should not affect runtime since the replaying only happens in case of a user-facing warning.

### CODE REVIEW

File: tactics/hints.ml [hunk 1]
```diff
@@ -916,19 +902,27 @@ let fresh_global_or_constr env sigma poly cr =
...
   let try_apply f =
-    try Some (f (c, cty, ctx)) with Failure _ -> None in
+    try
+      let (hd, hint) as ans = f (c, cty, ctx) in
+      let () = match hint.code.obj with
+      | ERes_pf _ -> if not eapply then failwith "make_apply_entry"
+      | _ -> ()
+      in
+      Some ans
+    with Failure _ -> None
```
Comment (SkySkimmer) [inline tactics/hints.ml:916]: Should we avoid raising just to be caught?
```suggestion
    try
      let (hd, hint) as ans = f (c, cty, ctx) in
      match hint.code.obj with
      | ERes_pf _ -> None
      | _ -> Some ans 
    with Failure _ -> None
```

Comment (ppedrot) [reply tactics/hints.ml:916]: I thought about that but the other part of the code still relies on raising exceptions, so I didn't want to change this behaviour too far...

Comment (ppedrot) [reply tactics/hints.ml:916]: (Your change is wrong btw, it needs to check eapply on the relevant branch.)

Comment (ppedrot) [reply tactics/hints.ml]: Done.

File: tactics/hints.ml [hunk 2]
```diff
@@ -916,19 +902,25 @@ let fresh_global_or_constr env sigma poly cr =
...
-let make_resolves env sigma flags info ~check ~poly ?name cr =
+let make_resolves env sigma (eapply, hnf) info ~check ~poly ?name cr =
   let c, ctx = fresh_global_or_constr env sigma poly cr in
   let cty = Retyping.get_type_of env sigma c in
   let try_apply f =
-    try Some (f (c, cty, ctx)) with Failure _ -> None in
+    try
+      let (_, hint) as ans = f (c, cty, ctx) in
+      match hint.code.obj with
+      | ERes_pf _ -> if not eapply then None else Some ans
```
Comment (SkySkimmer) [inline tactics/hints.ml:912]:
```suggestion
      | ERes_pf _ when not eapply -> None
```
?

Comment (ppedrot) [reply tactics/hints.ml:912]: I hate when clauses, they break linearity and are easy to mess up. Please don't force me to write one.

Comment (SkySkimmer) [reply tactics/hints.ml:912]: ok

**Comment by ppedrot** [comment]
@mattam82 would you assign?

**Review by mattam82 — APPROVED** [review]
That's a simple internal code refactoring, looking fine to me.

**Comment by mattam82** [comment]
@coqbot merge now

**Pull request merged**